xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-api/src/main/java/org/xwiki/notifications/filters/NotificationFilterManager.java

```diff
@@ -59,6 +60,23 @@
     Collection<NotificationFilter> getAllFilters(DocumentReference user, boolean onlyEnabled)
         throws NotificationException;
 
+    /**
+     * Get all notifications filters that are enabled to the given user and matching the requested filtering moment.
+     *
+     * @param user reference to the user
+     * @param onlyEnabled either or not only filters enabled for the user should be collected
+     * @param filteringMoments when the filter should be used
+     * @return the collection of notification filters enabled to the user.
+     * @throws NotificationException if an error happens
+     * @since 10.4RC1
+     */
+    default Collection<NotificationFilter> getAllFilters(DocumentReference user, boolean onlyEnabled,
+        Set<NotificationFilter.FilteringMoment> filteringMoments)
```

Comment:
I don't understand why we need a Set here. I feels like you would use this method to get filters for:
* pre filtering only
* post filtering only
* all

And this can be expressed with a single FilteringMoment parameter already.

Reply:
Because you usually want the filters for pre-filtering AND both, or post-filtering AND both. You can check how I used it in UserEventManager and DefaultNotificationParametersFactory.

Reply:
> Because you usually want the filters for pre-filtering AND both

Sure but my point is that it's what the parameter value pre-filtering should already mean. IMO it's just bad naming in the FilteringMoment enum, it should probably just be `PREFILTERING`, `POSTFILTERING` and `BOTH` (not even sure we really need BOTH actually since we could use null as "any phase" instead).

Reply:
> Sure but my point is that it's what the parameter value pre-filtering should already mean. 

I don't agree but in this case you would have two different meaning depending where the enum value is used: 
using `PREFILTERING` in `NotificationFilter#getPhase` would mean **only** `PREFILTERING` phase, while using `NotificationFilterManager#getAllFilters` with `PREFILTERING` parameter would return `PREFILTERING` **and** `BOTH` filters. 
I prefer if the phase names are explicit (hence the `ONLY_` prefix), as the API to use them, even if it makes it a bit more verbose to use.

Reply:
What about having NotificationFilter#isPhaseSupported instead of NotificationFilter#getPhase ?

Reply:
> What about having NotificationFilter#isPhaseSupported instead of NotificationFilter#getPhase ?

could be an elegant solution indeed. Not sure if we should keep `BOTH` to be explicit if you want to use `isPhaseSupported(BOTH)` or if we should just drop it.

Reply:
I don't think BOTH is needed in that configuration, you would just pass `null` as "I don't want to filter based on the phase, just give me everything".

Reply:
> I don't think BOTH is needed in that configuration

ok dropping it for now then

Reply:
Note that now the problem is that you cannot get the filters that are ONLY working on postfiltering, you always get all those supporting postfiltering  (meaning both + postfiltering)

Reply:
Which is not right I think actually... since it implies that you will call much filters than only the post-filtering ones when you don't need need to..

Reply:
So I just pushed the new solution, but I'm not entirely sure about it, see my comment here: https://github.com/xwiki/xwiki-platform/pull/1418/files#diff-2a2836c0e09aac5b088303fcf643a499R398